docs/design_notes/module_lifecycle.rst

```diff
@@ -0,0 +1,378 @@
+
+Definition
+=============
+
+Let's start with a high-level overview of the Module lifecycle. We start by defining a simple Module:
+
+
+.. testcode::
+
+  class MLP(nn.Module):
+    # 1. Attribute annotations
+    hidden_size: int
+    out_size: int
+
+    # 2. The ``setup`` method
+    def setup(self):
+      self.hidden = nn.Dense(self.hidden_size)
+      self.out = nn.Dense(self.out_size)
+
+    # 3. User methods  
+    def __call__(self, x):
+      a = self.hidden(x)
+      h = nn.relu(a)
+      return self.out(h)
+
+
+This Module consists of:
+
+#. **Attribute annotations**, defined as `dataclass <https://docs.python.org/3/library/dataclasses.html>`_ fields. These annotations automatically define a constructor.
+#. **The ``setup`` method**, which creates submodules and assigns them to attributes.
+#. **User methods**. By convention, most Modules have just one  ``__call__`` method, but you can define multiple methods or use different method names.
+
+Construction/initialization
+=============================
+
+Now we want to construct and use the ``MLP`` Module:
+
+
+.. testcode::
+
+  mlp = MLP(hidden_size=5, out_size=3)
+  x = jax.numpy.ones((1, 2))
+  variables = mlp.init(random.PRNGKey(0), x)
+  y = mlp.apply(variables, x)
+
+
+First, we construct an instance of ``MLP`` and pass the construction attributes. Note that construction here is different from what you might expect if you are not used to Functional Programming patterns. The ``MLP`` constructor does not actually create variables or any internal state whatsoever. It's best to think of it as a specification or template of the Module that contains functionality but no data.
+
+Let's first take a closer look at initialization. Surprisingly, there actually is no special initialization path in Flax. Calling ``init`` is just a special case of ``apply`` and we could also code it as:
+
+
+.. testcode:
+
+  # equivalent to: variables = mlp.init(random.PRNGKey(0), x)
+  _, variables = mlp.apply({}, x, rngs={"params": random.PRNGKey(0)}, mutable=True)
+
+
+Thus, ``init`` is nothing more than a special case of ``apply`` where:
+
+#. We call a Module without any initial variables (an empty dict).
+#. A PRNG generator named “params” is always passed for randomly initializing parameters (using the parameter initialization function).
+#. All variable collections are set to mutable (``mutable=True``). When a collection is mutable, existing variables can be updated and new variables can be created. Thus, inside ``init`` variables can be initialized in any variable collection and they are all added to the returned variable dictionary.
+
+Lifecycle
+=============
+
+
+Now that we understand ``init`` is just a special case of ``apply``, let's analyze what ``.apply(...)`` does in more detail. In fact, most of the complexity of Modules resides in the ``apply`` method. The "Module lifecycle" consists of constructing and ``apply``-ing a Module. We can summarize the Module lifecycle as follows:
```

Comment:
Nit: 

- "Special" [``init``] is repeated 3 or 4 times here (in the `apply` context), so maybe rephrase it a little.
- "We understand" is a presumption - maybe use "you learned about"

e.g.

```suggestion
Now that you have learned about ``init`` (a special case of ``apply``), let's analyze what ``.apply(...)`` does in more detail. In fact, most of the complexity of Modules resides in the ``apply`` method. The "Module lifecycle" consists of constructing and ``apply``ing a Module. We can summarize the Module lifecycle as follows:
```